tensorflow/lite/experimental/shlo/ops/convolution.cc

```diff
@@ -0,0 +1,1171 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/experimental/shlo/ops/convolution.h"
+
+#include <algorithm>
+#include <cstddef>
+#include <string>
+#include <type_traits>
+
+#include "absl/status/status.h"
+#include "tensorflow/lite/experimental/shlo/dispatch.h"
+#include "tensorflow/lite/experimental/shlo/ops/dot_general.h"
+#include "tensorflow/lite/experimental/shlo/ops/unary_elementwise.h"
+#include "tensorflow/lite/experimental/shlo/ops/util.h"
+#include "tensorflow/lite/experimental/shlo/tensor.h"
+
+namespace shlo_ref {
+
+template <class T>
+using DimVector = absl::InlinedVector<T, 6>;
+
+bool IsUnique(DimVector<int64_t>& vec) {
+  std::sort(vec.begin(), vec.end());
+  return std::unique(vec.begin(), vec.end()) == vec.end();
+}
+
+bool IsInRange(DimVector<int64_t>& vec, size_t N) {
+  for (int64_t dim : vec) {
+    if (dim >= N || dim < 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <DataType storage_type>
+absl::Status PrepareImpl(ConvolutionOp& op, const Tensor& lhs,
+                         const Tensor& rhs, Tensor& output) {
+  using StorageT = StorageType<storage_type>;
+
+  // Transpose prepare
+  const int64_t* window_spacial_pointer =
+      op.attributes.kernel_spacial_dimensions.GetDataAs<DataType::kSI64>();
+  const int64_t* output_spacial_pointer =
+      op.attributes.output_spacial_dimensions.GetDataAs<DataType::kSI64>();
+  const int64_t* input_spacial_pointer =
+      op.attributes.input_spacial_dimensions.GetDataAs<DataType::kSI64>();
+
+  std::vector<StorageT> lhs_permutation_values(
+      static_cast<int64_t>(lhs.Rank()));
+  lhs_permutation_values[0] = op.attributes.input_batch_dimension;
+  lhs_permutation_values[1] = op.attributes.input_feature_dimension;
+  DimVector<DimensionSize> lhs_shape_dims(lhs.Rank());
+  lhs_shape_dims[0] =
+      lhs.shape().Dim(static_cast<size_t>(op.attributes.input_batch_dimension));
+  lhs_shape_dims[1] = lhs.shape().Dim(
+      static_cast<size_t>(op.attributes.input_feature_dimension));
+  for (size_t i = 0; i < lhs.Rank() - 2; ++i) {
+    lhs_shape_dims[i + 2] =
+        lhs.shape().Dim(static_cast<size_t>(input_spacial_pointer[i]));
+    lhs_permutation_values[i + 2] = input_spacial_pointer[i];
+  }
+  // malloc is used to have the storage space available out of prepare function
+  // scope and it's pointer is stored in class data member to
+  // deallocate the memory in destructor.
```

Comment:
I'd prefer to figure out a container for storing these rather than using malloc directly. I made a utility class [TensorWithData](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/experimental/shlo/tensor_with_data.h) that could pair a Tensor with the backing data in a `std::vector<std::byte>`, but it was originally aimed at tests. Would something like that work? 

That said, do all of these need to be Tensors? The permutations & transposed seem to simply be 1d vectors.

Reply:
Alternatively, you could use a `std::vector<std::byte>` directly that would handle the memory for you in place of `malloc` calls.

Ideally, the ops should not own any memory and only keep a pointer to the external data that is provided. I understand that if you want to call another op you may need to create that op's parameters.

Reply:
changed to std::vector<std::byte>(you can check it in PR#67245)